backend-public/contracts/ryder-handles-controller.clar

```diff
@@ -0,0 +1,195 @@
+      ;; signature must be unused
+      (asserts! (is-none (map-get? renewal-signatures approval-signature)) err-signature-already-used) 
+      (map-set renewal-signatures approval-signature true)
+      (try! (pay-fees price (some namespace)))
+      (try! (contract-call? .community-handles name-renewal namespace name new-owner zonefile-hash))
+      (ok true)))
+
+(define-private (pay-fees (price uint) (namespace-or-escrow (optional (buff 20))))
+  (let ((amount-controller-admin (/ (* price u70) u100))
+        (amount-community (- price amount-controller-admin))
+        (controller-admin (var-get contract-owner))        
+        (community-treasury (match namespace-or-escrow
+                              namespace (default-to controller-admin (map-get? community-treasuries namespace))
+                              (as-contract tx-sender))))
+    (and (> amount-controller-admin u0)
+        (try! (stx-transfer? amount-controller-admin tx-sender controller-admin)))
+    (and (> amount-community u0)
+        (try! (stx-transfer? amount-community tx-sender community-treasury)))
+    (ok true)))
+
+(define-private (pay-fees-from-escrow (price uint) (namespace (buff 20)))
+    (let ((amount-controller-admin (/ (* price u70) u100))
+          (amount-community (- price amount-controller-admin))
+          (community-treasury (default-to (var-get contract-owner) (map-get? community-treasuries namespace))))
+        (and (> amount-community u0)
+            (try! (as-contract (stx-transfer? amount-community tx-sender community-treasury))))
+        (ok true)))
+
+;;
+;; admin functions
+;;
+(define-public (set-price (amount-in-ustx uint))
+    (begin
+        (try! (is-contract-owner))
+        (var-set price-in-ustx amount-in-ustx)
+        (ok true)))
+
+(define-public (set-community-treasury (namespace (buff 20)) (new-treasury principal))
+   (begin
+        (try! (is-contract-owner))
+        (map-set community-treasuries namespace new-treasury)
+        (ok true)))
+
+(define-public (set-approval-pubkey (new-pubkey (buff 33)))
+   (begin
+        (try! (is-contract-owner))
+        (var-set approval-pubkey new-pubkey)
+        (ok true)))
+
+
+(define-public (set-contract-owner (new-owner principal))
+    (begin
+        (try! (is-contract-owner))
+        (var-set contract-owner new-owner)
+        (ok true)))
+
+;; hand over control of namespace to new controller
+;; can only be called by contract owner of this contract
+(define-public (set-namespace-controller (namespace (buff 20)) (new-controller principal))
+    (begin
+        (try! (is-contract-owner))
+        (try! (as-contract (contract-call? .community-handles set-namespace-controller namespace new-controller)))
+        (ok true)))
+
+;;
+;; utilities
+;;
+
+;; @desc retrieve unused fees from escrow
+;; If name-register wasn't called successfully the community amount is in escrow
```

Comment:
Failed registrations happened quite a lot with .btc. Should we provide a pathway for registrants to reclaim their tokens directly?

Reply:
It would open up the names for spamming if it is too easy to reclaim tokens.

Reply:
Actually, it is not an issue and we could add that in the future.